packages/edit-site/src/components/navigate-to-link/index.js

```diff
@@ -28,14 +23,14 @@ export default function NavigateToLink( {
 				if ( success ) {
 					let newTemplateId = data.ID;
 					if ( newTemplateId === null ) {
-						const { getEntityRecord } = select( 'core' );
-						newTemplateId = templateIds
-							.map( ( id ) =>
-								getEntityRecord( 'postType', 'wp_template', id )
-							)
-							.find(
-								( template ) => template.slug === data.post_name
-							).id;
+						const { getEntityRecords } = select( 'core' );
+						newTemplateId = getEntityRecords(
+							'postType',
+							'wp_template',
+							{
+								slug: data.post_name,
+							}
+						)[ 0 ].id;
```

Comment:
The slug approach does not guarantee you'll get the same post ID that is loaded in the editor. It doesn't account for multiple auto-drafts, a published template, etc.

Was the previous approach making REST requests? All of those template IDs should be preloaded.

Reply:
> The slug approach does not guarantee you'll get the same post ID that is loaded in the editor. It doesn't account for multiple auto-drafts, a published template, etc.

Oh, okay. I think that's something we _should_ seek to resolve though -- possibly through some modifications to the endpoint, in order to reflect the [template loading mechanism](https://github.com/WordPress/gutenberg/blob/3b8965e197ba5bbcc24d9b8dfa1f759601416a52/lib/template-loader.php).

Aside, is the 'proper' resolution documented anywhere? I.e. when to use the published template, when to fall back to an auto-draft, when to generate one (?) etc? If it isn't, it'd be good to have it on file somewhere so it's easier to reason about.

It'd be good to have a client-side mechanism to request an up-to-date list of templates, so we can use the familiar `useSelect`/`useDispatch` pattern to update those. This will help solving glitches like [this](https://github.com/WordPress/gutenberg/pull/19141#issuecomment-589305471), and will overall help decouple the client from the PHP-provided variables (which in turn will eventually allow us to directly integrate `edit-site` with e.g. Calypso).

Reply:
>Oh, okay. I think that's something we should seek to resolve though -- possibly through some modifications to the endpoint, in order to reflect the template loading mechanism.

We shouldn't modify the standard REST API collection behavior. We can introduce a separate endpoint for finding the currently published template for a given slug.

>Aside, is the 'proper' resolution documented anywhere? I.e. when to use the published template, when to fall back to an auto-draft, when to generate one (?) etc? If it isn't, it'd be good to have it on file somewhere so it's easier to reason about.

Published templates are always used first and mean the template was customized. Auto drafts are only used when the templates haven't been customized. We should preferably use the latest ones, but because certain preloading flows can run template resolution more than once, you can end up in a situation where the site editor loaded the `N-1` auto drafts and then when you do this fetch by slug, you end up with an `N` auto draft.

This isn't documented because it's a regression I recently found when catching up with the preloading work.

>It'd be good to have a client-side mechanism to request an up-to-date list of templates, so we can use the familiar useSelect/useDispatch pattern to update those. This will help solving glitches like this, and will overall help decouple the client from the PHP-provided variables (which in turn will eventually allow us to directly integrate edit-site with e.g. Calypso).

Agreed.

Reply:
> > Oh, okay. I think that's something we should seek to resolve though -- possibly through some modifications to the endpoint, in order to reflect the template loading mechanism.
> 
> We shouldn't modify the standard REST API collection behavior. We can introduce a separate endpoint for finding the currently published template for a given slug.

Agreed. If the required query can still be done on the client side (and doesn't impact perf too negatively), we can also resort to that. (I didn't mean to mess with the collection endpoint format, I was rather considering adding some standard modifications akin to e.g. your #21851.)

> > Aside, is the 'proper' resolution documented anywhere? I.e. when to use the published template, when to fall back to an auto-draft, when to generate one (?) etc? If it isn't, it'd be good to have it on file somewhere so it's easier to reason about.
> 
> Published templates are always used first and mean the template was customized. Auto drafts are only used when the templates haven't been customized. We should preferably use the latest ones, but because certain preloading flows can run template resolution more than once, you can end up in a situation where the site editor loaded the `N-1` auto drafts and then when you do this fetch by slug, you end up with an `N` auto draft.

Thanks for explaining, that's very helpful!

> This isn't documented because it's a regression I recently found when catching up with the preloading work.

Also good to know, thanks :+1:

Reply:
Perhaps, a new top-level query param would be nice. That way, we can keep using `core/data` semantically in the client, but leave the awkward publish/auto-draft logic in the server. Something like `?loaded` or `?resolved`.

Reply:
BTW, I just realized that the logic you described seems to be implemented in [`useTemplatePartPost`](https://github.com/WordPress/gutenberg/blob/caf52c16304e375e8a28047f47d3b69c81c9aa68/packages/block-library/src/template-part/edit/use-template-part-post.js#L6), right? So it might make sense to use this in `NavigateToLink` in order to centralize the implementation of this behavior, and allow us to change it if we end up making changes to the endpoint.

> Perhaps, a new top-level query param would be nice. That way, we can keep using `core/data` semantically in the client, but leave the awkward publish/auto-draft logic in the server. Something like `?loaded` or `?resolved`.

Yeah, keeping that logic on the server side sounds like a good idea. Would you add those query args to the existing endpoint (using `rest_wp_template_part_collection_params` and `rest_wp_template_part_query`)?

Reply:
>BTW, I just realized that the logic you described seems to be implemented in useTemplatePartPost, right? So it might make sense to use this in NavigateToLink in order to centralize the implementation of this behavior, and allow us to change it if we end up making changes to the endpoint.

Yes

>Yeah, keeping that logic on the server side sounds like a good idea. Would you add those query args to the existing endpoint (using rest_wp_template_part_collection_params and rest_wp_template_part_query)?

Yes

Would you like to do those things in this PR instead?

Reply:
Yeah, I'll give that a shot tomorrow!

Reply:
Thanks!

Reply:
Prep PRs took a while :sweat_smile: WIP PR for this feature: https://github.com/WordPress/gutenberg/pull/21981